### PR TITLE
Add `Duration::new_{checked,unchecked}` constructors

### PR DESCRIPTION
**Pull request by jmillikin**
`Duration::new_checked()` returns `None` in cases where `Duration::new()` would panic.

`Duration::new_unchecked()` requires that the provided nanosecond count be less than 1 billion (the number of nanoseconds in a second).

Tracking issue: https://github.com/rust-lang/rust/issues/125748

### CODE REVIEW

**Comment by jhpratt** [comment]
Per https://github.com/rust-lang/libs-team/issues/117#issuecomment-2135749768, only `Duration::new_checked` was approved. The implementation of this looks good; I will approve if `Duration::new_unchecked` is removed.

**Comment by jmillikin** [comment]
I have no need for `new_checked()` myself, and only added it because there was some interest from the reviewer.

I'll keep this PR open until @the8472 can weigh in on the assembly listings in the ACP. If there's truly no interest from the Rust devs regarding `new_unchecked` then I'll close it.

File: library/core/src/time.rs [hunk 1]
```diff
@@ -223,6 +223,68 @@ impl Duration {
...
+    /// assert_eq!(Duration::new_checked(5, 0), Some(Duration::new(5, 0)));
+    /// assert_eq!(Duration::new_checked(u64::MAX, u32::MAX), None);
+    /// ```
+    #[unstable(feature = "duration_new_checked", issue = "125748")]
+    #[inline]
+    #[must_use]
+    pub const fn new_checked(secs: u64, nanos: u32) -> Option<Duration> {
+        if nanos < NANOS_PER_SEC {
+            // SAFETY: nanos < NANOS_PER_SEC, therefore nanos is within the valid range
+            Some(unsafe { Duration::new_unchecked(secs, nanos) })
```
Comment (dev-ardi) [inline library/core/src/time.rs:250]: Better construct the duration here inline

File: library/core/src/time.rs [hunk 2]
```diff
@@ -223,6 +223,68 @@ impl Duration {
...
+        } else {
+            let secs = match secs.checked_add((nanos / NANOS_PER_SEC) as u64) {
+                Some(secs) => secs,
+                None => return None,
+            };
+            let subsec_nanos = nanos % NANOS_PER_SEC;
+            // SAFETY: nanos % NANOS_PER_SEC < NANOS_PER_SEC, therefore
+            // subsec_nanos is within the valid range.
+            Some(unsafe { Duration::new_unchecked(secs, subsec_nanos) })
```
Comment (dev-ardi) [inline library/core/src/time.rs:259]: ditto

File: library/core/src/time.rs (hunk 1 above)
Comment (jmillikin) [reply library/core/src/time.rs]: Done.

File: library/core/src/time.rs (hunk 2 above)
Comment (jmillikin) [reply library/core/src/time.rs]: Done.

**Pull request closed without merging**